src/c_string.rs

```diff
@@ -0,0 +1,468 @@
+    pub fn as_c_str(&self) -> &CStr {
+        if cfg!(debug_assertions) {
+            // When in debug builds, ensure
+            CStr::from_bytes_with_nul(&self.vec).unwrap()
+        } else {
+            unsafe { CStr::from_bytes_with_nul_unchecked(&self.vec) }
+        }
+    }
+
+    /// How many bytes were inserted to this [`CString`] so far, considering its
+    /// null terminator.
+    ///
+    /// Must always be bigger than zero, since even an empty [`CString`]
+    /// ends in a zero byte.
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    /// use std::ffi::{c_char, CStr};
+    ///
+    /// // Length is one (null terminator only), capacity is 10
+    /// let mut cstr = CString::<11>::new();
+    ///
+    /// // Insert 5 bytes to it
+    /// cstr.push_bytes(b"/etc/").unwrap();
+    /// // Length is 6 (5 bytes inserted plus null terminator)
+    /// assert_eq!(cstr.len(), 6);
+    /// cstr.push_bytes(b"dconf").unwrap();
+    ///
+    /// assert_eq!(cstr.to_str(), Ok("/etc/dconf"));
+    /// ```
+    pub fn len(&self) -> usize {
+        self.as_bytes_with_nul().len()
+    }
+
+    /// Calculates the length of `self.vec` would have if it appended `bytes`.
+    fn size_if_appended_bytes(&self, bytes: &[u8]) -> Option<usize> {
+        match bytes.last() {
+            None => None,
+            Some(0) if bytes.len() < 2 => None,
+            Some(0) => {
+                // `bytes` is null-terminated and so is `self.vec`.
+                // Adding up both would account for 2 null bytes when only a single byte
+                // would end up in the resulting CString.
+                Some(self.vec.len() + bytes.len() - 1)
+            }
+            Some(_) => {
+                // No terminating null byte in `bytes` but there's one in
+                // `self.vec`, so the math lines up nicely.
+                //
+                // In the case that `bytes` has a null byte anywhere else, we would
+                // error after `memchr` is called. So there's no problem.
+                Some(self.vec.len() + bytes.len())
+            }
+        }
+    }
+
+    // pub fn push(&mut self, byte: u8) -> Result<(), ()> {
+
+    // }
+
+    /// Extends the [`CString`] with the given bytes.
+    ///
+    /// This method fails if the [`CString`] would not have enough capacity to append the bytes or
+    /// if the bytes contain an interior null byte (a zero byte not at the buffer's final position).
+    ///
+    /// If there's a null byte at the end, however, the function does not fail.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut cstr = CString::<10>::new();
+    ///
+    /// cstr.push_bytes(b"hey").unwrap();
+    /// cstr.push_bytes(b" there\0").unwrap();
+    ///
+    /// assert_eq!(cstr.to_str(), Ok("hey there"));
+    /// ```
+    pub fn push_bytes(&mut self, bytes: &[u8]) -> Result<(), CapacityError> {
```

Comment:
Should we rename this `extend_from_bytes` or `extend_from_slice`?

Reply:
I renamed it `extend_from_bytes` for now. `extend_from_slice` is a bit ambiguous because it might mean either byte or `CStr` slices.